### PR TITLE
Issue #4639: added suppport for lambdas in UnnecessaryParentheses

### PR DESCRIPTION
**Pull request by rnveach**
Issue #4639

Regression to come.

### CODE REVIEW

**Comment by rnveach** [comment]
@romani 1 test should fail.
Do we want to add new token to default and should we add it to our configuration?

**Comment by rnveach** [comment]
PR only regression on new token: http://rveach.no-ip.org/checkstyle/regression/reports/99/

http://rveach.no-ip.org/checkstyle/regression/reports/99/openjdk9/index.html#A24
http://rveach.no-ip.org/checkstyle/regression/reports/99/openjdk9/xref/openjdk9/src/java.base/share/classes/java/util/function/Function.java.html#L88
Removing the parentheses from the lambda makes it uncompilable. I will add this to the inputs and prevent it from being flagged.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesLambdas.java [hunk 1]
```diff
@@ -0,0 +1,19 @@
+package com.puppycrawl.tools.checkstyle.checks.coding.unnecessaryparentheses;
+
+public class InputUnnecessaryParenthesesLambdas {
+    int foo(int y) {
+        MathOperation addition = (x) -> x + x;
+        MathOperation addition2 = (x) -> { return x + x; };
+        MathOperation addition3 = x -> x + x;
+        MathOperation2 addition4 = (a, b) -> a + b;
```
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesLambdas.java]: please make a case like `() -> ....` and `(A a, B b) -> .....`
there are even more interesting cases like `t1 -> t2 -> apply(t1, t2);` (see it [there](http://rveach.no-ip.org/checkstyle/regression/reports/99/Vavr/xref/Vavr/vavr/src-gen/main/java/io/vavr/CheckedFunction2.java.html#L160))
We need more cases of singleline-nested lambdas, please review vavr project, they love lambdas.

Comment (rnveach) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesLambdas.java]: > `t1 -> t2 -> apply(t1, t2);`

I am not able to make a copy of this and need some help.

>  singleline-nested lambdas

Do you mean lambdas that are nested inside expressions?

Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesLambdas.java]: > Do you mean lambdas that are nested inside expressions?

inside Lambdas and probably few cases for inside expressions , just to make sure we do make difference between them.

**Comment by rnveach** [comment]
> it make sense to put new token to defaults.

Done.

> We need more cases 

I added what I could but I need help adding anymore as I am not very good with lambdas.

**Comment by romani** [comment]
> I need help adding anymore as I am not very good with lambdas.

I took this from vavr code, compiled on my side:
```
    interface Lambda<R> extends Serializable {
        Lambda<?> curried();
    }

    public interface Function1<T1, R> extends Lambda<R>, Function<T1, R> {
        @Override
        default Function1<T1, R> curried() {
            return this;
        }
    }

    public interface CheckedFunction1<T1, R> extends Lambda<R> {
        R apply(T1 t1) throws Throwable;

        @Override
        default CheckedFunction1<T1, R> curried() {
            return this;
        }
    }

    public interface CheckedFunction2<T1, T2, R> extends Lambda<R> {

        R apply(T1 t1, T2 t2) throws Throwable;

        default CheckedFunction1<T2, R> apply(T1 t1) {
            return (T2 t2) -> apply(t1, t2);
        }
        @Override
        default Function1<T1, CheckedFunction1<T2, R>> curried() {
                    return t1 -> t2 -> apply(t1, t2);
        }
    }
```

**Comment by rnveach** [comment]
@romani Thanks very much. I was able to incorporate it in.
I will generate new regression reports.

**Comment by rnveach** [comment]
New regression: http://rveach.no-ip.org/checkstyle/regression/reports/100/

**Comment by romani** [comment]
shippable is relaunched.

**Comment by rnveach** [comment]
shippable passed.

**Comment by romani** [comment]
@rnveach , I think it will be good to make a mode to "do not ignore Excludes" it will help for such mode of report.
Such exceptions blow mind .....

**Pull request merged**